operator/internal/handlers/internal/storage/secrets.go

```diff
@@ -443,13 +460,35 @@ func extractS3ConfigSecret(s *corev1.Secret, credentialMode lokiv1.CredentialMod
 		// In the STS case region is not an optional field
 		if len(region) == 0 {
 			return nil, fmt.Errorf("%w: %s", errSecretMissingField, storage.KeyAWSRegion)
+		} else if len(endpoint) != 0 && err != nil {
+			return nil, err
 		}
 		return cfg, nil
 	default:
 		return nil, fmt.Errorf("%w: %s", errSecretUnknownCredentialMode, credentialMode)
 	}
 }
 
+func validateS3Endpoint(endpoint string, region string) error {
+	parsedURL, err := url.Parse(endpoint)
+	if err != nil {
+		return errS3EndpointUnparseable
+	}
+	if parsedURL.Scheme != "http" && parsedURL.Scheme != "https" {
+		return errS3EndpointUnsuportedScheme
+	}
+	if !strings.HasSuffix(endpoint, awsEndpoint) {
+		return errS3EndpointNotAWS
+	}
+	if !strings.HasSuffix(endpoint, fmt.Sprintf("%s.%s", region, awsEndpoint)) {
+		return errS3EndpointWrongRegion
+	}
+	if !strings.HasSuffix(endpoint, fmt.Sprintf("s3.%s.%s", region, awsEndpoint)) {
+		return errS3EndpointNotS3
+	}
```

Comment:
This effectively removes the possibility to use non-AWS S3 endpoints, which is not what this change is supposed to do. We want to provide some additional validations for when the user _wants to use AWS S3_. If they are using something else, then we do not apply these validations, but we also do not forbid to use it.

```golang
const (
  awsEndpointSuffix      = ".amazonaws.com"
)

var (
  errS3EndpointAWSInvalid       = errors.New("endpoint for AWS S3 must include correct region"
)
```

```suggestion
	if strings.HasSuffix(endpoint, awsEndpointSuffix) {
		validEndpoint := fmt.Sprintf("https://s3.%s%s", region, awsEndpointSuffix)
		if endpoint != validEndpoint {
			return fmt.Errorf("%w: %s", errS3EndpointAWSInvalid, validEndpoint)
		}
	}
```